src/drivers/cmakefileapi/api_helpers.ts

```diff
@@ -14,19 +14,23 @@ import { fs } from '@cmt/pr';
 import * as path from 'path';
 import * as nls from 'vscode-nls';
 import rollbar from '@cmt/rollbar';
+import * as pLimit from 'p-limit';
 
 nls.config({ messageFormat: nls.MessageFormat.bundle, bundleFormat: nls.BundleFormat.standalone })();
 const localize: nls.LocalizeFunc = nls.loadMessageBundle();
 
 const log = logging.createLogger('cmakefileapi-parser');
 
+// Limits the concurrent async access to the file system to avoid errors such as "EMFILE: too many open files".
+const fsAccessLimiter = pLimit(50);
+
 export async function createQueryFileForApi(api_path: string): Promise<string> {
     const query_path = path.join(api_path, 'query', 'client-vscode');
     const query_file_path = path.join(query_path, 'query.json');
     const requests = { requests: [{ kind: 'cache', version: 2 }, { kind: 'codemodel', version: 2 }, { kind: 'toolchains', version: 1 }] };
     try {
-        await fs.mkdir_p(query_path);
-        await fs.writeFile(query_file_path, JSON.stringify(requests));
+        await fsAccessLimiter(() => fs.mkdir_p(query_path));
```

Comment:
Since `fs` is actually our wrapper around NodeJS's `fs` would it make more sense to make this change in `pr.ts` instead?  Then all callers of the filesystem APIs will get this benefit and we won't have to remember to import this module everywhere.

Reply:
If that is what you want, I can apply it to `pr.ts`. Does 50 sound like a sensible value to you or should we increase it?

Reply:
My biggest concern is around `readdir`.  I'm not sure we need to limit file/directory enumeration, isn't it just file reads that are the problem?  Or does the OS hold on to file descriptors for the other operations too?

Reply:
But then again, in https://www.npmjs.com/package/rimraf docs, it seems to mention that `readdir` uses file handles, so I don't know what to think anymore.

Reply:
@bobbrow 

I did more extensive analysis for linux side at least:
* `fs.exists` - uses `fs.tryStat` - does not open FD
* `fs.existsSync` - not async, do not care
* `fs.readFile` - opens FD via `uv_fs_open`
* `fs.writeFile` - opens FD via `uv_fs_open`
* `fs.readdir` - uses `uv_fs_scandir` (yes, scandir) - does not open FD
* `fs.mkdir` - uses `uv_fs_mkdir` - does not open FD
* `fs.mkdtemp` - uses `uv_fs_mkdtemp` - does not open FD
* `fs.rename` - uses `uv_fs_rename` - does not open FD
* `fs.stat` - uses `uv_fs_stat` - does not open FD
* `fs.walk` - uses `fs.stat` and `fs.readdir` - does not open FD
* `fs.tryStat` - uses `fs.stat` - does not open FD
* `fs.readlink` - uses `uv_fs_readlink` - does not open FD
* `fs.unlink` - uses `uv_fs_unlink` - does not open FD
* `fs.appendFile` - uses `fs.writeFile` - opens FD via `uv_fs_open`
* `fs.mkdir_p` - uses `fs.exists`, `fs.stat` and `fs.mkdir` - does not open FD
* `fs.copyFile` - opens FD via `uv_fs_open`
* `fs.rmdir` - found no evidence that FD is opened, but the docs claim that it does it. At least it is handled by rimraf then, no need for a limit.

I will update the PR to limit only the functions which I'm 100% opens the FD.

Reply:
Awesome analysis. Thank you for being so thorough!

Reply:
Would you like me to squish the commits?

Reply:
That won't be necessary.  We will do a squash merge.  Thanks for asking though.